src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs

```diff
@@ -284,87 +295,162 @@ public IEnumerable<WorkloadDownload> GetDownloads(IEnumerable<WorkloadId> worklo
+                    }
+                }
+            }
 
-                    if (!unneededBandRecordPaths.Any())
+            // Garbage collect workload packs
+            var featureBandsWithWorkloadInstallationRecords = _installationRecordRepository.GetFeatureBandsWithInstallationRecords();
+
+            Dictionary<(WorkloadPackId packId, string packVersion), List<SdkFeatureBand>> packInstallRecords = GetAllPackInstallRecords();
+            foreach (var (packId, packVersion) in packInstallRecords.Keys)
+            {
+                var referencingFeatureBands = packInstallRecords[(packId, packVersion)];
+                List<SdkFeatureBand> featureBandsToRemove = new();
+                foreach (var referencingFeatureBand in referencingFeatureBands)
+                {
+                    if (cleanAllPacks)
                     {
-                        continue;
+                        //  Remove pack record if we're cleaning all packs
+                        featureBandsToRemove.Add(referencingFeatureBand);
                     }
-
-                    // Save the pack info in case we need to delete the pack
-                    var jsonPackInfo = File.ReadAllText(unneededBandRecordPaths.First());
-                    foreach (var unneededPackRecord in unneededBandRecordPaths)
+                    else if (!installedSdkFeatureBands.Contains(referencingFeatureBand) ||
+                             !featureBandsWithWorkloadInstallationRecords.Contains(referencingFeatureBand))
                     {
-                        File.Delete(unneededPackRecord);
+                        //  Remove pack record if corresponding feature band is not installed, or does not have any workloads installed
+                        featureBandsToRemove.Add(referencingFeatureBand);
                     }
-
-                    // If there are no pack records left in the directory we garbage collected, we need to delete the pack and its director(ies).
-                    if (!bandPackRecordPaths.Except(unneededBandRecordPaths).Any())
+                    else if (referencingFeatureBand.Equals(_sdkFeatureBand) && !garbageCollector.PacksToKeep.Contains((packId, packVersion)))
                     {
-                        Directory.Delete(packVersionDir);
-                        var deletablePack = GetPackInfo(packVersionDir);
-                        if (deletablePack == null)
-                        {
-                            // Pack no longer exists in manifests, get pack info from installation record
-                            deletablePack = JsonSerializer.Deserialize(jsonPackInfo, typeof(PackInfo)) as PackInfo;
-                        }
-                        DeletePack(deletablePack);
+                        //  For current feature band, garbage collect packs that the garbage collector didn't mark as ones to keep
+                        featureBandsToRemove.Add(referencingFeatureBand);
                     }
```

Comment:
Consider combining all 3 predicates into 1 if-statement with `||` since they all do the same thing.

Reply:
This is to try to make it easier to read and understand.  I think this way is clearer than one really long expression.